src/linker/Linker/DocumentationSignatureParser.cs

```diff
@@ -0,0 +1,666 @@
+						// Instead of constructing a representation of the array bounds, we
+						// use the original input to represent the bounds, and later match it
+						// against the generated strings for types in signatures.
+						// This ensures that we will only resolve members with supported array bounds.
+						typeReference += id.Substring (boundsStartIndex, boundsEndIndex - boundsStartIndex);
+						continue;
+					}
+
+					if (PeekNextChar (id, index) == '*') {
+						index++;
+						typeReference += '*';
+						continue;
+					}
+
+					break;
+				}
+
+				if (PeekNextChar (id, index) == '@') {
+					index++;
+					typeReference += '@';
+				}
+
+				results[i] = typeReference;
+				endIndex = index;
+			}
+
+			index = endIndex;
+		}
+
+		private static void ParseTypeParameterSymbol (string id, ref int index, IGenericParameterProvider? typeParameterContext, List<string> results)
+		{
+			// skip the first `
+			Debug.Assert (PeekNextChar (id, index) == '`');
+			index++;
+
+			Debug.Assert (
+				typeParameterContext == null ||
+				(typeParameterContext is MethodDefinition && typeParameterContext.GenericParameterType == GenericParameterType.Method) ||
+				(typeParameterContext is TypeDefinition && typeParameterContext.GenericParameterType == GenericParameterType.Type)
+			);
+
+			if (PeekNextChar (id, index) == '`') {
+				// `` means this is a method type parameter
+				index++;
+				var methodTypeParameterIndex = ReadNextInteger (id, ref index);
+
+				if (typeParameterContext is MethodDefinition methodContext) {
+					var count = methodContext.HasGenericParameters ? methodContext.GenericParameters.Count : 0;
+					if (count > 0 && methodTypeParameterIndex < count) {
+						results.Add ($"``{methodTypeParameterIndex}");
+					}
+				}
+			} else {
+				// regular type parameter
+				var typeParameterIndex = ReadNextInteger (id, ref index);
+
+				var typeContext = typeParameterContext is MethodDefinition methodContext
+					? methodContext.DeclaringType
+					: typeParameterContext as TypeDefinition;
+
+				if (typeParameterIndex >= 0 ||
+					typeParameterIndex < typeContext?.GenericParameters.Count) {
+					// No need to look at declaring types like Roslyn, because type parameters are redeclared.
+					results.Add ("`" + typeParameterIndex);
+				}
+			}
+		}
+
+		private static void ParseNamedTypeSymbol (string id, ref int index, IGenericParameterProvider? typeParameterContext, List<string> results)
+		{
+			Debug.Assert (results.Count == 0);
+			var nameBuilder = new StringBuilder ();
+			// loop for dotted names
+			while (true) {
+				var name = ParseName (id, ref index);
+				if (String.IsNullOrEmpty (name))
+					return;
+
+				nameBuilder.Append (name);
+
+				List<string>? typeArguments = null;
+				int arity = 0;
+
+				// type arguments
+				if (PeekNextChar (id, index) == '{') {
+					typeArguments = new List<string> ();
+					if (!ParseTypeArguments (id, ref index, typeParameterContext, typeArguments)) {
+						continue;
```

Comment:
Should we output a warning in places where the parsing fails?

Reply:
Ideally yes, but I'm not sure in this case. Roslyn doesn't raise a warning when this is malformed in `SuppressMessage` - it simply ignores it in that case. So I think linker should do the same in this case.

Reply:
Agreed it would be nice. Currently the way it will be done is that the caller of the parser gets to decide what kind of warnings/errors to log - which unfortunately doesn't distinguish between syntax errors and resolution errors. I initially tried to separate those two cases, but ended up just following the Roslyn implementation.